src/main/java/com/azure/cosmos/samples/distributedbulk/DocumentBulkExecutor.java

```diff
@@ -0,0 +1,117 @@
+    private final Function<T, String> idExtractor;
+
+    public DocumentBulkExecutor(
+        CosmosAsyncContainer cosmosAsyncContainer,
+        Function<T, String> idExtractor) {
+
+        Objects.requireNonNull(cosmosAsyncContainer, "Argument 'cosmosAsyncContainer' must not be null.");
+        Objects.requireNonNull(idExtractor, "Argument 'idExtractor' must not be null.");
+        this.cosmosAsyncContainer = cosmosAsyncContainer;
+        this.idExtractor = idExtractor;
+    }
+
+    /*
+    public BulkDeleteResponse deleteAll(List<Pair<String, String>> pkIdPairsToDelete) {
+
+    }
+
+    public BulkImportResponse importAll(Collection<?> documents) {
+    }*/
+
+    /**
+     * Upserts all documents in bulk mode
+     * @param documents a collection of documents for a logical batch
+     * @return A BulkImportResponse with aggregated diagnostics and eventual failures.
+     */
+    public BulkImportResponse upsertAll(Stream<T> documents) {
+        return this.upsertAll(
+            documents,
+            new DocumentBulkExecutorOperationStatus());
+    }
+
+    /**
+     * Upserts all documents in bulk mode
+     * @param documents a collection of documents for a logical batch
+     * @param status can be used to track the status of the bulk ingestion for this batch at real-time
+     * @return A BulkImportResponse with aggregated diagnostics and eventual failures.
+     */
+    public BulkImportResponse upsertAll(
+        Stream<T> documents,
+        DocumentBulkExecutorOperationStatus status) {
+
+        Objects.requireNonNull(documents, "Argument 'documents' must not be null.");
+        Objects.requireNonNull(status, "Argument 'status' must not be null.");
+
+        try {
+            Stream<CosmosItemOperation> docsToUpsert = documents
+                .map(doc -> {
+                        String id = this.idExtractor.apply(doc);
+                        return CosmosBulkOperations.getUpsertItemOperation(
+                                doc,
+                                new PartitionKey(id),
+                                null,
+                                new OperationContext(id, status.getOperationId()));
+                    });
+            return executeBulkOperations(docsToUpsert, status);
+        } catch (Throwable t) {
+            logger.error("Failed to upsert documents - {}", t.getMessage(), t);
+            throw new IllegalStateException("Failed to upsert documents", t);
+        }
+    }
+
+    private BulkImportResponse executeBulkOperations(
+        Stream<CosmosItemOperation> operations,
+        DocumentBulkExecutorOperationStatus status) {
+
+        Instant startTime = Instant.now();
+        try (BulkWriter writer = new BulkWriter(
+            this.cosmosAsyncContainer,
+            status)) {
+
+            operations
+                .forEach(writer::scheduleWrite);
+
+            logger.info("All items of batch {} scheduled.", status.getOperationId());
+            writer.flush();
```

Comment:
Here `flush()` is essentially like a latch which waits for a Stream<CosmosItemOperation> to be completed?

Reply:
flush will signal to the incoming flux of operations that no more operations will be scheduled. And if BulkExecutor has a batch not yet reaching its max batch size it will be send to backend imemdiately vs. waiting for the flush interval.